#### Foundations of Machine Learning Aug-Nov 2024
##### Sakshi Badole CS24MTECH11008
Question 4: __Soft Margin SVM__ for handwritten digits recognition ***Only for classes 1(+1) and 5(-1)*** 

##### Import Statements


##### Import Statements
Libraries used :
* SVC from sklearn.svm 
* StandardScaler for data proprocessing
* pandas
* numpy

In [1]:
from sklearn.svm import SVC
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler

##### Read data into python dataframes

##### Read data into pandas dataframe
read.csv reads train and test data into respective pandas dataframes 

In [2]:
Train_data = pd.read_csv('http://www.amlbook.com/data/zip/features.train', delim_whitespace=True, header=None)
Test_data = pd.read_csv('http://www.amlbook.com/data/zip/features.test', delim_whitespace=True, header=None)


C:\Users\sakshi badole\AppData\Local\Temp\ipykernel_14620\3920857288.py:1: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  Train_data = pd.read_csv('http://www.amlbook.com/data/zip/features.train', delim_whitespace=True, header=None)
C:\Users\sakshi badole\AppData\Local\Temp\ipykernel_14620\3920857288.py:2: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  Test_data = pd.read_csv('http://www.amlbook.com/data/zip/features.test', delim_whitespace=True, header=None)


##### Taking only those rows which have label either 1 or 5

In [3]:
Train_data = Train_data.loc[(Train_data.iloc[:, 0] == 1) | (Train_data.iloc[:, 0] == 5)]
Test_data = Test_data.loc[(Test_data.iloc[:, 0] == 1) | (Test_data.iloc[:, 0] == 5)]

##### Updating labels :  '1' becomes '+1' and '5' becomes '-1' 

In [4]:
Train_data.loc[Train_data.iloc[:, 0] == 1, 0] = +1
Train_data.loc[Train_data.iloc[:, 0] == 5, 0] = -1
Test_data.loc[Test_data.iloc[:, 0] == 1, 0] = +1
Test_data.loc[Test_data.iloc[:, 0] == 5, 0] = -1

##### Dividing the train and test data into features and labels using iloc as first column of train and test dataframes
Using .values to convert dataframes into Numpy ndarray

In [5]:
X_train = Train_data.iloc[:,1:].values
X_test = Test_data.iloc[:,1:].values
y_train = Train_data.iloc[:,0].values
y_test = Test_data.iloc[:,0].values


##### Data preprocessing : standardization

In [6]:
X_train = StandardScaler().fit_transform(X_train)
X_test = StandardScaler().fit_transform(X_test)

##### Part a) Linear kernel with Soft margin SVM

In [7]:
svm_linear = SVC(kernel='linear', C=0.1)

In [8]:
svm_linear.fit(X_train, y_train)

SVC(C=0.1, kernel='linear')

In [9]:
print("The test accuracy for svm linear kernel is:",svm_linear.score(X_test, y_test))

The test accuracy for svm linear kernel is: 0.9834905660377359


In [10]:
print("Number of Support vectors for the svm linear kernel is: ",svm_linear.support_vectors_.shape[0])


Number of Support vectors for the svm linear kernel is:  80


***
##### Part b) Using the first {50,100, 200, 800} data points

(b).1 For first 50 data points

Creates new train data and labels by extracting the **first 50 data points** from the train data and labels

In [11]:
X_train_b = X_train[0:50]
y_train_b = y_train[0:50]
print(X_train_b.shape)
print(y_train_b.shape)

(50, 2)
(50,)


In [12]:
svm_linear.fit(X_train_b, y_train_b)

SVC(C=0.1, kernel='linear')

In [13]:
print("The test accuracy for svm linear kernel for the first 50 data points",svm_linear.score(X_test, y_test))

The test accuracy for svm linear kernel for the first 50 data points 0.9575471698113207


In [14]:
print("Number of Support vectors for the first 50 data points are",svm_linear.support_vectors_.shape[0])


Number of Support vectors for the first 50 data points are 14


***
(b).2 For first 100 data points

Creates new train data and labels by extracting the **first 100 data points** from the train data and labels

In [15]:
X_train_b = X_train[0:100]
y_train_b = y_train[0:100]
print(X_train_b.shape)
print(y_train_b.shape)

(100, 2)
(100,)


In [16]:
svm_linear.fit(X_train_b, y_train_b)

SVC(C=0.1, kernel='linear')

In [17]:
print("The test accuracy for svm linear kernel for the first 100 data points",svm_linear.score(X_test, y_test))

The test accuracy for svm linear kernel for the first 100 data points 0.9764150943396226


In [18]:
print("Number of Support vectors for the first 100 data points are",svm_linear.support_vectors_.shape[0])


Number of Support vectors for the first 100 data points are 21


***
(b).3 For first 200 data points

Creates new train data and labels by extracting the **first 200 data points** from the train data and labels

In [19]:
X_train_b = X_train[0:200]
y_train_b = y_train[0:200]
print(X_train_b.shape)
print(y_train_b.shape)

(200, 2)
(200,)


In [20]:
svm_linear.fit(X_train_b, y_train_b)

SVC(C=0.1, kernel='linear')

In [21]:
print("The test accuracy for svm linear kernel for the first 200 data points",svm_linear.score(X_test, y_test))

The test accuracy for svm linear kernel for the first 200 data points 0.9811320754716981


In [22]:
print("Number of Support vectors for the first 200 data points are",svm_linear.support_vectors_.shape[0])


Number of Support vectors for the first 200 data points are 30


***
(b).4 For first 800 data points

Creates new train data and labels by extracting the **first 800 data points** from the train data and labels

In [23]:
X_train_b = X_train[0:800]
y_train_b = y_train[0:800]
print(X_train_b.shape)
print(y_train_b.shape)

(800, 2)
(800,)


In [24]:
svm_linear.fit(X_train_b, y_train_b)

SVC(C=0.1, kernel='linear')

In [25]:
print("The test accuracy for svm linear kernel for the first 800 data points",svm_linear.score(X_test, y_test))

The test accuracy for svm linear kernel for the first 800 data points 0.9834905660377359


In [26]:
print("Number of Support vectors for the first 800 data points are",svm_linear.support_vectors_.shape[0])


Number of Support vectors for the first 800 data points are 56


__Part (b) Analysis of SVM with linear kernel over different number of data points__

|No. of Data Points |Test Accuracy|Number of Support Vectors|
|:-----|:----:|----:|
|50 |0.957547 |14 |
|100 |0.976415 |21 |
|200 |0.981132 |30 |
|800 |0.983490 |56 |

***
##### Part c) Polynomial kernel comparing Q=2 and Q=5 (Q means the degree of the polynomial)

i) When C = 0.0001, training error is higher at Q = 5. __FALSE__

In [27]:
svm_polynomial = SVC(kernel='poly', C=0.0001, degree=2)
svm_polynomial.fit(X_train, y_train)
print("Training Error when Q=2 at C=0.0001 is",1-svm_polynomial.score(X_train, y_train))


Training Error when Q=2 at C=0.0001 is 0.356181934657271


In [28]:
svm_polynomial = SVC(kernel='poly', C=0.0001, degree=5)
svm_polynomial.fit(X_train, y_train)
print("Training Error when Q=5 at C=0.0001 is",1-svm_polynomial.score(X_train, y_train))


Training Error when Q=5 at C=0.0001 is 0.24791800128122998


ii. When C = 0.001, the number of support vectors is lower at Q = 5. __TRUE__

In [29]:
svm_polynomial = SVC(kernel='poly', C=0.001, degree=2)
svm_polynomial.fit(X_train, y_train)
print("Number of Support vectors when Q=2 and C=0.001 are",svm_polynomial.support_vectors_.shape[0])

Number of Support vectors when Q=2 and C=0.001 are 1088


In [30]:
svm_polynomial = SVC(kernel='poly', C=0.001, degree=5)
svm_polynomial.fit(X_train, y_train)
print("Number of Support vectors when Q=5 and C=0.001 are",svm_polynomial.support_vectors_.shape[0])

Number of Support vectors when Q=5 and C=0.001 are 687


iii. When C = 0.01, training error is higher at Q = 5. __FALSE__

In [31]:
svm_polynomial = SVC(kernel='poly', C=0.01, degree=2)
svm_polynomial.fit(X_train, y_train)
print("Training Error when Q=2 and C=0.01 is",1-svm_polynomial.score(X_train, y_train))


Training Error when Q=2 and C=0.01 is 0.16463805253042918


In [32]:
svm_polynomial = SVC(kernel='poly', C=0.01, degree=5)
svm_polynomial.fit(X_train, y_train)
print("Training Error when Q=5 and C=0.01 is",1-svm_polynomial.score(X_train, y_train))


Training Error when Q=5 and C=0.01 is 0.12812299807815508


iv. When C = 1, test error is lower at Q = 5. __TRUE__

In [33]:
svm_polynomial = SVC(kernel='poly', C=1, degree=2)
svm_polynomial.fit(X_train, y_train)
print("Test Error when Q=2 and C=1 is",1-svm_polynomial.score(X_test, y_test))


Test Error when Q=2 and C=1 is 0.14622641509433965


In [34]:
svm_polynomial = SVC(kernel='poly', C=1, degree=5)
svm_polynomial.fit(X_train, y_train)
print("Test Error when Q=5 and C=1 is",1-svm_polynomial.score(X_test, y_test))


Test Error when Q=5 and C=1 is 0.09669811320754718


***
##### Part d) Radial Basis Kernel in soft-margin SVM approach

Train and Test errors for C = 0.01

In [35]:
svm_rbf = SVC(kernel='rbf', C=0.01)
svm_rbf.fit(X_train, y_train)
print("Training Error when C=0.01 is", 1- svm_rbf.score(X_train, y_train))
print("Test Error when C=0.01 is", 1-svm_rbf.score(X_test, y_test))

Training Error when C=0.01 is 0.00832799487508007
Test Error when C=0.01 is 0.02358490566037741


Train and Test errors for C = 1

In [36]:
svm_rbf = SVC(kernel='rbf', C=1)
svm_rbf.fit(X_train, y_train)
print("Training Error when C=1 is",1-svm_rbf.score(X_train, y_train))
print("Test Error when C=1 is",1-svm_rbf.score(X_test, y_test))


Training Error when C=1 is 0.0032030749519538215
Test Error when C=1 is 0.018867924528301883


Train and Test errors for C = 100

In [37]:
svm_rbf = SVC(kernel='rbf', C=100)
svm_rbf.fit(X_train, y_train)
print("Training Error when C=100 is",1-svm_rbf.score(X_train, y_train))
print("Test Error when C=100 is",1-svm_rbf.score(X_test, y_test))


Training Error when C=100 is 0.0032030749519538215
Test Error when C=100 is 0.01650943396226412


Train and Test errors for C = 10^4

In [38]:
svm_rbf = SVC(kernel='rbf', C=10**4)
svm_rbf.fit(X_train, y_train)
print("Training Error when C=10^4 is",1-svm_rbf.score(X_train, y_train))
print("Test Error when C=10^4 is",1-svm_rbf.score(X_test, y_test))


Training Error when C=10^4 is

 0.002562459961563124
Test Error when C=10^4 is 0.018867924528301883


Train and Test errors for C = 10^6

In [39]:
svm_rbf = SVC(kernel='rbf', C=10**6)
svm_rbf.fit(X_train, y_train)
print("Training Error when C=10^6 is",1-svm_rbf.score(X_train, y_train))
print("Test Error when C=10^6 is",1-svm_rbf.score(X_test, y_test))


Training Error when C=10^6 is 0.0
Test Error when C=10^6 is 0.021226415094339646


__Part (d) Analysis of Soft-Margin SVM with gaussian(rbf) kernel over different values of C__

|Value of C |Training Error|Test Error|
|:-----|:----:|:----:|
|0.01 |0.00832799487508007 |0.02358490566037741 |
|1 |0.0032030749519538215 |0.018867924528301883 |
|100 |0.0032030749519538215 |0.01650943396226412 |
|10^4 |0.002562459961563124 |0.018867924528301883 |
|10^6 |0.0 |0.021226415094339646 |

The lowest training error is for __C = 10^6__ and lowest test error is for __C=100__